pyomo/core/expr/visitor.py

```diff
@@ -1181,7 +1181,7 @@ def visit(self, node):
             return
 
         # TODO: Confirm that this has the right semantics
-        if not node.is_variable_type() and node.is_fixed():
+        if not node.is_variable_type() and not node.is_constant():
```

Comment:
I think you still need the `is_fixed` test to make sure you don't catch other nodes types like Expression objects.  That is,
```python
if not node.is_variable_type() and node.is_fixed() and not node.is_constant():
```

Reply:
I was thinking about this too. I'm certainly fine with leaving the `is_fixed` check, but doesn't the `xbfs_yield_leaves` method already filter out expressions?

Reply:
Also, is it possible to have expression objects for which is_fixed returns True? I'm thinking about the NPV expressions.

Reply:
You are correct: some part of the walker does correctly handle things like `Expression` and `MonomialTermExpression`.

Yes: `is_fixed()` should return correct values for any expression object (it triggers a walk of the expression tree).  That said, NPV objects have `is_constant()` True by decree (no walk required - in fact, as I recall, NPV nodes were added explicitly to speed up the `is_constant` test).

Reply:
I think the `LinearExpression` issue points to a bigger problem, which is that the `args` and `nargs()` properties/methods do not return anything meaningful for `LinearExpression`.

Reply:
So we also need to make sure that the `identify_mutable_parameters` function does not return NPV expressions:

```
In [1]: import pyomo.environ as pe

In [2]: m = pe.ConcreteModel()

In [3]: m.p = pe.Param(mutable=True)

In [4]: e = 2*m.p

In [5]: e
Out[5]: <pyomo.core.expr.numeric_expr.NPV_ProductExpression at 0x7ff1becfcb10>

In [6]: e.is_constant()
Out[6]: False

In [7]: e.is_fixed()
Out[7]: True
```

Reply:
I just re-added the `is_fixed()` check, but I can't think of a way to test it. That is, I can't think of a non-variable leaf node that has both `is_constant()` False and `is_fixed()` False.

Reply:
> So we also need to make sure that the `identify_mutable_parameters` function does not return NPV expressions

Is it possible for an NPV expression to be leaf node?

Reply:
No - I don't think so.  Apart from `Expression` and `LinearExpression`, no expression-like objects should be leaf nodes.

Reply:
Sorry - I apparently can't keep it in my head that this uses `xbfs_yield_leaves`.